add rfc for migrating to dotnet-script from scriptcs

IsaacCalligeros95 opened the pull request:
# About

This is a post RFC about migrating from scriptcs to dotnet-script for executing C# scripts within a run a script step.

# Release

No, this is meant to gather feedback about the proposed changes, particularly from customers using this functionality on older systems.

# Blog image idea

I have used the existing Octopus salesperson at laptop with headset and icons representing customer feedback image. seen [here](https://octopus.com/blog/rfc-second-ecs-integration-with-octopus).

# Pre-requisites

- [x] The draft is complete and this post is ready to be reviewed.
- [x] I'm familiar with the [writing for Octopus TL;DR](https://www.octopus.design/writing/writing-tldr/).
- [x] The PR build passes validation, and if it doesn't, I've checked the [common validations errors](https://style.octopus.com/writing-at-octopus-tldr#common-validation-errors) and none of those apply.
- [x] The images I've included follow the [rules for working with images](https://www.octopus.design/writing/working-with-images/).
- [ ] I've added alt text for all the images I've included (alt text describes image to people unable to see it, 125 characters max)
- [ ] One screenshot (if any are included) is suitable for Twitter. This screenshot should be 16:9 ratio and at least 1200 x 675 pixels. For screenshots from the Octopus Web Portal, please use Chrome and the browser’s zoom function to remove as much white space as possible. If screenshots are annotated (e.g., with boxes or arrows), please also provide a version that is unannotated for use on Twitter. Do not use dark mode for the Octopus Web Portal.

## How to review this PR
<!-- If there's anything you'd like reviewers to focus on, mention it here. -->
Initially, I'm planning on sending this out to some team members to verify technical details. After that all writing feedback would be greatly appreciated!

## Publication date
From our side as soon as it's ready, but happy to go with your suggestions.

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+We've received [user feedback](https://help.octopus.com/t/consider-use-dotnet-script-vs-scriptcs/22144) and [voting](https://octopusdeploy.uservoice.com/forums/170787-general/suggestions/31454668-allow-the-use-of-c-script-csx-using-net-core) requesting an update of the Calamari C# script executor from [scriptcs](https://github.com/scriptcs/scriptcs) to [dotnet-script](https://github.com/filipw/do ...
+
+This request for comments aims to gather feedback on the demand for this functionality and raise awareness of the tradeoffs in deprecating scriptcs. C# Scripting accounts for ~5% of our script steps and most instances of Calamari run on netcore 3.1 meaning this change is likely to only affect a small subset of customers.  The deployment targets affected by this change are linux targets using SSH  ...
```
```suggestion
This request for comments aims to gather feedback on the demand for this functionality and raise awareness of the tradeoffs in moving to dotnet-script, and deprecating scriptcs. C# Scripting account for ~5% of our script steps, so we want to ensure we understand the impact this change could have on our users.  

If you ar using C# scripts in your deployment processes, and are deploying to Linux targets using SSH and Mono, or to Windows tentacle targets running Windows versions earlier than 2012R2, the proposed changes could impact you.
```

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+This request for comments aims to gather feedback on the demand for this functionality and raise awareness of the tradeoffs in deprecating scriptcs. C# Scripting accounts for ~5% of our script steps and most instances of Calamari run on netcore 3.1 meaning this change is likely to only affect a small subset of customers.  The deployment targets affected by this change are linux targets using SSH  ...
+
+## How we propose to support established dotnet-script
```
Is this worded correctly?

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+## How we propose to support established dotnet-script
+
+This Request for Comments (RFC) proposes removing scriptcs in favour of dotnet-script. In doing so C# scripting will no longer be available  on Linux deployment targets using SSH, this can be resolved by using the self-contained Calamari which runs on netcore3.1. (Note check this final statement)
```
```suggestion
This RFC proposes removing `scriptcs` in favour of `dotnet-script`. 

One of the tradeoffs of this change is that C# scripting will no longer be available on Linux deployment targets using SSH with Mono. If you wish to run C# scripts against your SSH linux targets, you will need to reconfigure your SSH targets to use the self-contained Calamari which runs via netcore3.1. Targets using the Linux tentacle will continue to work as they always have. (Note check this final statement)
```

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+This Request for Comments (RFC) proposes removing scriptcs in favour of dotnet-script. In doing so C# scripting will no longer be available  on Linux deployment targets using SSH, this can be resolved by using the self-contained Calamari which runs on netcore3.1. (Note check this final statement)
+
+Calamari currently supports .Net Framework 4.0.0, 4.5.2 and Core 3.1 while dotnet-script only supports Core 3.1 and above. This means C# scripting will also be unavailable to deployments working with Tentacles installed on versions of Windows earlier than 2012R2, these run the Framework builds of Calamari. 
```
```suggestion
The other tradeoff we would make with this change is that `dotnet-script` only works with dotnet core 3.1 and above. This would mean C# scripting will be unavailable to deployments against Windows Tentacles installed on versions of Windows earlier than 2012 R2, as these run Full .NET Framework builds of Calamari. 
```

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+## How we propose to support established dotnet-script
+
+This Request for Comments (RFC) proposes removing scriptcs in favour of dotnet-script. In doing so C# scripting will no longer be available  on Linux deployment targets using SSH, this can be resolved by using the self-contained Calamari which runs on netcore3.1. (Note check this final statement)
```
I feel like we need to set a little more technical context here - the fact that Calamari is the thing that executes deployment scripts, and is the host for any C# scripts that execute during deployment.

andrewabest replied:
Nice , that is great!

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+| Feature                               | scriptcs          | dotnet-script    |
+|---------------------------------------|-------------------|------------------|
+| C# Version                            | 5                 | 8                |
+| Removes Mono Depdency                 | ❌                | ✅              |
```
```suggestion
| Removes Mono Depedency for Linux                 | ❌                | ✅              |
```

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
````diff
@@ -0,0 +1,88 @@
+Console.WriteLine(response.Content);
+```
+
+dotnet-script supports .Net 5 and 6, this should help towards future upgrades of Calamari and allow the use of the latest language features in your scripts.
````
I feel like this is a bit of a duplicate of the c#8 feature paragraph above.

IsaacCalligeros95 replied:
Removed

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+## When will this be released?
+
+We are still considering how many existing deployment processes this change is likely to affect. The reliability of your deployments are our main priority and as such we don't have an immediate plan to release these changes.
```
```suggestion
We are still evaluating how many users this change is likely to affect. We will not make or release the proposed changes until we have a clear picture of who we will impact, and what actions those who are impacted would need to take.
```

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+## We want your feedback
+
+We're still planning the second milestone, so now is a great time to help shape this new feature with your feedback. We have created a [GitHub issue to capture the discussion](https://github.com/OctopusDeploy/StepsFeedback/issues/7).
```
Need to adjust this paragraph.

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+## Conclusion
+
+In summary, the migration from ScriptCS to dotnet script includes results in the following changes:
```
```suggestion
In summary, the migration from `ScriptCS` to `dotnet-script` includes results in the following changes:
```

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+
+- C# scripting deprecated for Linux SSH targets running Mono
+- C# scripting deprecated for Windows deployment targets running on versions earlier than 2012R2
+- Support for language features up to C# 8 from C# 5
```
```suggestion
- Increase support for language features from C# 5 to C# 8
```

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,88 @@
+- C# scripting deprecated for Windows deployment targets running on versions earlier than 2012R2
+- Support for language features up to C# 8 from C# 5
+- Direct imports of Nuget packages in scripts
+- Removal of Mono dependence for calling scripts
```
```suggestion
- Removal of Mono requirement for running C# scripts on Linux targets
```

andrewabest reviewed (changes requested):
I've left a bunch of suggestions and changes. Another thing worth noting is just to be consistent with naming and casing for things like `ScriptCS` :) good first draft!

IsaacCalligeros95 commented:
> I've left a bunch of suggestions and changes. Another thing worth noting is just to be consistent with naming and casing for things like `ScriptCS` :) good first draft!

Thanks for all the feedback! I've added the changes and added a couple of sentences for technical context on Calamari.

mjhilton commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,93 @@
+
+This RFC proposes removing `ScriptCS` in favour of `dotnet-script`.
+
+To deploy software to your server we use [Tentacle](https://github.com/OctopusDeploy/OctopusTentacle), a lightweight service responsible for communicating with Octopus Server, and invoking [Calamari](https://github.com/OctopusDeploy/Calamari). Calamari is a command-line tool which knows how to perform the deployment, and is the host process for all deployment actions including script execution.
```
Eg:

> ## How we propose to support dotnet-script
> This RFC proposes... _etc_
>
> To deploy software.. _etc_
>
> `ScriptCS` is an old app built on the full .NET Framework, so in order to execute it on Linux targets, we require you to have Mono installed... _etc_
>
> `dotnet-script` is a modern implementation of C# scripting, built on `dotnet`, and so can run on all targets that support `dotnet` apps. The earliest Windows version that supports `dotnet-script` is Windows Server 2012 R2... _etc_
>
> ### Impacts we anticipate
> * C# scripting would no longer be available on Linux deployment targets configured to use SSH with Mono. If you wish to... _mitigation, etc_
> * C# scripting would no longer be available to deployments against Windows Tentacles installed on versions of Windows earlier than 2012 R2, as these run Full .NET Framework builds of Calamari. If you wish to... _mitigation, etc_

IsaacCalligeros95 replied:
👍 Sounds good, looking at it again this added context on the two tools should really tie it all together. I'll update in a moment, just trying to work through an implementation detail and want to understand it first 🙂

IsaacCalligeros95 replied:
I've made some changes along these lines, let me know what you think 🙂

mjhilton commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,93 @@
+
+## We want your feedback
+
+We're still considering this change, so now is a great time to help shape this proposal with your feedback. We have created a [GitHub issue to capture the discussion](https://github.com/OctopusDeploy/StepsFeedback/issues/7).
```
GitHub issue here would need to be changed before publishing

mjhilton commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,93 @@
+
+Thanks for reading this RFC. We hope you're as excited about the proposed migration to `dotnet-script` as we are.
+
+Any [feedback](https://github.com/OctopusDeploy/StepsFeedback/issues/7) you have is greatly appreciated.
```
and here

mjhilton reviewed (commented):
This post is great overall. Much more succinct than I would have written, which I think is a good thing :P 

I've highlighted one main part where I think some more explanation would help provide both clarity and better structure.

mjhilton commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -20,13 +20,19 @@ This request for comments (RFC) aims to gather feedback on the demand for this f
+To deploy software to your server we use [Tentacle](https://github.com/OctopusDeploy/OctopusTentacle), a lightweight service responsible for communicating with Octopus Server, and invoking [Calamari](https://github.com/OctopusDeploy/Calamari). Calamari is a command-line tool which knows how to perform the deployment, and is the host process for all deployment actions including script execution. W ...
 
-One of the tradeoffs of this change is that C# scripting will no longer be available on Linux deployment targets using SSH with Mono. If you wish to run C# scripts against your SSH linux targets, you will need to reconfigure your SSH targets to use the self-contained Calamari which runs via netcore3.1. Targets using the Linux tentacle will continue to work as they always have.
+Historically Calamari required Mono to be installed on your Linux targets to execute `ScriptCS` as it's built on the full .NET Framework. With the introduction of cross platform dotnet apps with netcore3.1 Linux can now natively run dotnet apps removing the complexity and overhead of Mono. Linux targets will receive the netcore3.1 Calamari by default with the exception of [Linux SSH targets](http ...
```
Just a note on tense here - we're making a proposal which describes a _potential_ state. 

When talking about the changes we propose to make, we should use a conditional future tense - because it's not yet certain that we're going to do it. Definite future is something like "\<y> will happen". Conditional future is "If we did \<x>, \<y> would happen."

For example, what you've got right now is a definite future tense:

> Linux targets **will** receive the netcore3.1 Calamari

The "will" here can be read in different ways. I'd recommend to use "would" instead. This makes a much clearer distinction between what is current state and what we propose to change. It makes quite a difference just to change the "will" to a "would:

> Linux targets **would** receive the netcore3.1 Calamari

or, to be even more explicit, you could say:

> **If we were to** implement this change, Linux targets **would** receive the netcore3.1 Calamari

mjhilton commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -20,13 +20,19 @@ This request for comments (RFC) aims to gather feedback on the demand for this f
-One of the tradeoffs of this change is that C# scripting will no longer be available on Linux deployment targets using SSH with Mono. If you wish to run C# scripts against your SSH linux targets, you will need to reconfigure your SSH targets to use the self-contained Calamari which runs via netcore3.1. Targets using the Linux tentacle will continue to work as they always have.
+Historically Calamari required Mono to be installed on your Linux targets to execute `ScriptCS` as it's built on the full .NET Framework. With the introduction of cross platform dotnet apps with netcore3.1 Linux can now natively run dotnet apps removing the complexity and overhead of Mono. Linux targets will receive the netcore3.1 Calamari by default with the exception of [Linux SSH targets](http ...
+
+`dotnet-script` is a modern implementation of C# scripting, built on dotnet, it can run on all targets that support dotnet apps (netcore3.1 and newer). This means Windows server versions will need to support dotnet apps to execute `dotnet-script`. Windows Server 2012 R2 only supports the full .Net Framework, meaning Windows Server 2012 R2 and earlier versions will no longer support C# scripting.
```
Likewise here, I find the use of "will" a little confusing. Let's be really explicit about what is potential future: 

> This means Windows server versions will need to support dotnet apps to execute `dotnet-script`. Windows Server 2012 R2 only supports the full .Net Framework, meaning Windows Server 2012 R2 and earlier versions will no longer support C# scripting.

could be much more clearly expressed as

> *If we made this change*, it *would* mean that C# scripts would only be able to be run on targets that support `dotnet`. Windows Server 2012 R2 and earlier only support .NET Framework, so these targets *would* lose the ability to run C# scripts.

mjhilton commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -20,13 +20,19 @@ This request for comments (RFC) aims to gather feedback on the demand for this f
+
+## Impacts
+
+One of the tradeoffs of this change is that C# scripting will no longer be available on Linux deployment targets using SSH with Mono. If you wish to run C# scripts against your SSH linux targets, you will need to reconfigure your SSH targets to use the self-contained Calamari which runs via netcore3.1. To do this select the Self-Contained Calamari target runtime on your SSH target, a guide can be ...
```
The definite future vs conditional future stuff here is less important because of the context (Impacts). But if you wanted to, you could reword it with similar principles:

> One of the tradeoffs of this change would be that C# scripting would no longer be available... 
> If you wished to continue to run C# scripts against your SSH Linux targets, you would need to...

etc :)

andrewabest commented:
This is looking really good. I think the biggest risk with the proposed move is C# scripts not working with netframework targets, so will be interesting what we hear about those scenarios. 

One thing we haven't called out in here that I think might be worth it is how impacted people might migrate their scripts so that they still work. "Upgrade your OS" is likely off the table, but it feels reasonable that people could change their script step to a powershell script step, and invoke a C# script that way. They just need to ensure scriptcs is installed on the host machine - they could do this via another script step, or possibly even by sending scriptCS as a package dependency of the script step and invoking it from there. Might be worth a quick spike - this would allow people to understand how they would dig themselves out of the incompatibility hole.

IsaacCalligeros95 commented:
> One thing we haven't called out in here that I think might be worth it is how impacted people might migrate their scripts so that they still work. "Upgrade your OS" is likely off the table, but it feels reasonable that people could change their script step to a powershell script step, and invoke a C# script that way. They just need to ensure scriptcs is installed on the host machine - they could do this via another script step, or possibly even by sending scriptCS as a package dependency of the script step and invoking it from there. Might be worth a quick spike - this would allow people to understand how they would dig themselves out of the incompatibility hole.

I've had a quick look at spiking this and have the following approach which is functional but limited in handling variable substitution within the ScriptContent. I'm still trying to see if there's a way to avoid having to refactor the script to pass Octopus variables as parameters but I'm just curious if you think including this approach is good enough?

`$ScriptContent = @"
#r "bin/Octopus.Client.dll"

using System;
using System.IO;
using Octopus.Client;

Console.Write(Env.ScriptArgs[0]);
Console.WriteLine(typeof(OctopusServerEndpoint));
"@

New-Item -Path . -Name "ScriptFile.csx" -ItemType "file" -Value $ScriptContent

scriptcs ScriptFile.csx -- $OctopusParameters["Octopus.Deployment.Id"]`

I'll also include an equivalent for Linux and how to install ScriptCS on those targets/from within a script.

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,132 @@
+
+The other tradeoff we would make with this change is that `dotnet-script` only works with netcore3.1 and above. This would mean C# scripting will be unavailable to deployments against Windows Tentacles installed on versions of Windows earlier than 2012 R2, as these run Full .NET Framework builds of Calamari. 
+
+## Workarounds
```
Nice! I think I'd prefer this section to be called "Migration" rather than "Workarounds" though.

andrewabest commented on blog/2022-q3/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,132 @@
+With these changes paramaters used in the C# script must be passed in through the ScriptCS arguments and referenced by Env.ScriptArgs[Index] inside the ScriptContent. 
+:::
+
+### Added functionality
```
It feels like this sectino should be above the migrations section, within the H2 "Impacts" section.

IsaacCalligeros95 replied:
I've moved this up, along with the Benefits of the proposed approach, I think they fit together. Matt and I had a bit of a re-review before I saw this feedback and we rewrote the (Workarounds/Migration) section.

andrewabest reviewed (approved):
I've made a couple of small suggestions, but overall this looks excellent Isaac!

mjhilton reviewed (approved):
LVGTM :)

TeganAli commented on blog/2022-07/migrate-from-scriptcs-to-dotnet-script/index.md:
```diff
@@ -0,0 +1,131 @@
+description: Provide your feedback on the proposed migration from scriptcs to dotnet-script.
+author: isaac.calligeros@octopus.com
+visibility: public
+published: 2022-07-21-1400
```
```suggestion
published: 2022-07-20-1400
```
